dali/kernels/scratch.h

```diff
@@ -0,0 +1,95 @@
+// Copyright (c) 2017-2018, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_KERNELS_SCRATCH_H_
+#define DALI_KERNELS_SCRATCH_H_
+
+#include <array>
+#include <utility>
+#include <type_traits>
+#include "dali/kernels/context.h"
+
+namespace dali {
+namespace kernels {
+
+template <typename T>
+class BumpAllocator {
+  static_assert(
+    std::is_pod<T>::value,
+    "BumpAllocator must be used with POD types");
+
+ public:
+  BumpAllocator() = default;
+  BumpAllocator(T *mem, size_t total) : memory_(mem), total_(total), used_(0) {}
+  BumpAllocator(BumpAllocator &&a) {
+    *this = std::move(a);
+  }
+
+  ~BumpAllocator() {
+    for (size_t i = 0; i < used_; i++) {
+      memory_[i].~T();
```

Comment:
don't really like invoking the destructors implicitly but I don't have a better suggestion. Anyone?

Reply:
I think this is this special, custom case we can do it. After all you are usually not developing custom allocators on your own.